pkg/util/debouncer/debouncer.go

```diff
@@ -0,0 +1,279 @@
+package debouncer
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/instrument"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	ErrBufferFull = errors.New("debouncer buffer full")
+)
+
+type KeyFunc[T any] func(T) string
+type ProcessFunc[T any] func(context.Context, T) error
+
+type metrics struct {
+	itemsAddedCounter           prometheus.Counter
+	itemsDroppedCounter         prometheus.Counter
+	itemsProcessedCounter       prometheus.Counter
+	processingErrorsCounter     prometheus.Counter
+	processingDurationHistogram prometheus.Histogram
+}
+
+func newMetrics(reg prometheus.Registerer, name string) *metrics {
+	return &metrics{
+		itemsAddedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: name + "_debouncer_items_added_total",
+			Help: "Total number of items added to the debouncer",
+		}),
+		itemsDroppedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: name + "_debouncer_items_dropped_total",
+			Help: "Total number of items dropped due to a full buffer",
+		}),
+		itemsProcessedCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: name + "_debouncer_items_processed_total",
+			Help: "Total number of items processed by the debouncer",
+		}),
+		processingErrorsCounter: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: name + "_debouncer_processing_errors_total",
+			Help: "Total number of errors during processing",
+		}),
+		processingDurationHistogram: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
+			Name:                            name + "_debouncer_processing_duration_seconds",
+			Help:                            "Time taken to process items",
+			Buckets:                         instrument.DefBuckets,
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  160,
+			NativeHistogramMinResetDuration: time.Hour,
+		}),
+	}
+}
+
+type DebouncerOpts[T any] struct {
+	BufferSize   int
+	KeyFunc      KeyFunc[T]
+	MinWait      time.Duration
+	MaxWait      time.Duration
+	Reg          prometheus.Registerer
+	Name         string
+	ErrorHandler func(T, error)
+}
+
+type Debouncer[T any] struct {
+	opts   DebouncerOpts[T]
+	buffer chan T
+
+	// mutex protecting the debouncers map.
+	debouncersMu sync.Mutex
+	debouncers   map[string]*keyDebouncer[T]
```

Comment:
I find the naming here a bit confusing. We have a `Debouncer` with a `debouncers` field. A keyDebouncer ensures an item with a specific key (for example "default:core:dashboard") doesn't get processed too often. The Debouncer type is more of a collection of such debouncers. Don't have a naming suggestion right now unfortunately.

Reply:
We can call the main struct "Group" and the keys perhaps "Debouncer"?

Reply:
Changed 👍